### PR TITLE
feat: DB plugin multiplexing

### PR DESCRIPTION
**Pull request by fairclothjm**
# Description

This PR makes use of the https://github.com/hashicorp/vault/tree/feat-mux feature branch and is merging in the current state of the DB plugin multiplexing work. 

Currently the following phases have been completed:
- Move plugin process management to Vault Core
- Multiplex core process management

### CODE REVIEW

File: sdk/database/dbplugin/v5/plugin_server.go [hunk 1]
```diff
@@ -31,7 +31,36 @@ func ServeConfig(db Database) *plugin.ServeConfig {
...
-		HandshakeConfig:  handshakeConfig,
+		HandshakeConfig:  HandshakeConfig,
+		VersionedPlugins: pluginSets,
+		GRPCServer:       plugin.DefaultGRPCServer,
+	}
+
+	return conf
+}
+
+func ServeMultiplex(factory func() (Database, error)) {
```
Comment (fairclothjm) [inline sdk/database/dbplugin/v5/plugin_server.go:42]: I am wondering if this file can be merged into main in a separate PR since the Oracle database plugin will need the updated SDK?

File: vault/plugin_catalog.go [hunk 2]
```diff
@@ -59,12 +84,98 @@ func (c *Core) setupPluginCatalog(ctx context.Context) error {
...
+	c.multiplexedClients[pluginName] = mpc
+	c.logger.Debug("set the MultiplexedClient for", "pluginName", pluginName)
+
+	return mpc
+}
+
+// getPluginClient returns a client for managing the lifecycle of a plugin
+// process and an ID for a newly created entry in the MultiplexedClients'
+// connections map.
+func (c *PluginCatalog) getPluginClient(ctx context.Context, sys pluginutil.RunnerUtil, pluginRunner *pluginutil.PluginRunner, namedLogger log.Logger, isMetadataMode bool) (plugin.ClientProtocol, string, error) {
```
Comment (briankassouf) [inline vault/plugin_catalog.go:135]: I'm not sure we need to pass sys or pluginRunner into this function, our Core object should be able to reference or derive these itself. Maybe it just needs the plugin name we want to create?

File: vault/plugin_catalog.go [hunk 3]
```diff
@@ -59,12 +84,98 @@ func (c *Core) setupPluginCatalog(ctx context.Context) error {
...
+	}
+
+	mpc := c.getMultiplexedClient(pluginRunner.Name)
+
+	if mpc.client == nil {
+		c.logger.Debug("spawning a new plugin process")
+		client, err := pluginRunner.RunConfig(ctx,
+			pluginutil.Runner(sys),
+			pluginutil.PluginSets(v5.PluginSets),
+			pluginutil.HandshakeConfig(v5.HandshakeConfig),
```
Comment (briankassouf) [inline vault/plugin_catalog.go:148]: These should probably be passed into this function instead of assuming we are generating a database plugin.

Perhaps a new arg type:
```
PluginClientConfig{
    Name string
    PluginSets ..
    HandshakeConfig ...
}
```

File: vault/plugin_catalog.go (hunk 2 above)
Comment (fairclothjm) [reply vault/plugin_catalog.go:135]: Thanks! I will take a closer look

File: vault/plugin_catalog.go (hunk 3 above)
Comment (fairclothjm) [reply vault/plugin_catalog.go:148]: Thanks! Yes, I had the same thought and I will consider the arg type

File: sdk/database/dbplugin/v5/grpc_server.go [hunk 4]
```diff
@@ -16,19 +18,91 @@ var _ proto.DatabaseServer = gRPCServer{}
...
+	return nil, fmt.Errorf("no database instance found")
 }
 
 // Initialize the database plugin
 func (g gRPCServer) Initialize(ctx context.Context, request *proto.InitializeRequest) (*proto.InitializeResponse, error) {
+	impl, err := g.getOrCreateDatabase(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if impl == nil {
```
Comment (kalafut) [inline sdk/database/dbplugin/v5/grpc_server.go:94]: When would this be hit?

File: sdk/database/dbplugin/v5/grpc_server.go [hunk 5]
```diff
@@ -16,19 +18,91 @@ var _ proto.DatabaseServer = gRPCServer{}
...
+		return nil, err
+	}
+
+	g.instances[id] = db
+
+	return db, nil
+}
+
+func (g gRPCServer) getDatabase(ctx context.Context) (Database, error) {
+	g.Lock()
```
Comment (kalafut) [inline sdk/database/dbplugin/v5/grpc_server.go:69]: Can this be an `RLock`?

File: vault/plugin_catalog.go [hunk 6]
```diff
@@ -59,12 +84,98 @@ func (c *Core) setupPluginCatalog(ctx context.Context) error {
 	return nil
 }
 
+func (c *PluginCatalog) removeMultiplexedClient(ctx context.Context, name, id string) {
+	if _, ok := c.multiplexedClients[name]; !ok {
```
Comment (kalafut) [inline vault/plugin_catalog.go:88]: Do we need any locking on `c.multiplexedClients`? (I've just started looking at the code and maybe the locks, if needed, are elsewhere.)

Comment (fairclothjm) [reply vault/plugin_catalog.go:88]: Yes, it most likely will need locks. I haven't added any around this object yet.

File: vault/plugin_catalog.go (hunk 2 above)
Comment (fairclothjm) [reply vault/plugin_catalog.go:135]: I have looked into this for a while today. I am not sure we can omit pluginRunner since this function is also in the path of registering plugins and we haven't yet written the plugin entry to storage.

File: vault/plugin_catalog.go (hunk 6 above)
Comment (fairclothjm) [reply vault/plugin_catalog.go:88]: Thinking about this a bit more, I think the lock for PluginCatalog should be sufficient. So I don't think any locking is needed on c.multiplexedClients.

File: plugins/database/postgresql/postgresql.go [hunk 7]
```diff
@@ -48,7 +48,13 @@ var (
 	singleQuotedPhrases = regexp.MustCompile(`('.*?')`)
 )
 
-// New implements builtinplugins.BuiltinFactory
+func NewWithMultiplex() (dbplugin.Database, error) {
```
Comment (calvn) [inline plugins/database/postgresql/postgresql.go:51]: I wonder if `NewWithMultiplex` is needed since it's pretty much the same as `New`. From the plugin's perspective, all they might to do to enable multiplexing is for their main func to call `ServeMultiplex` with the factory func provided (which is what you're doing in `postgresql-database-plugin/main.go`).

I noticed that we do something that's pretty similar already on today's non-DB plugins: https://github.com/hashicorp/vault/blob/a24653cc5cca475962b27b71dcba29a4aa94d6ee/sdk/plugin/serve.go#L22-L30

File: sdk/database/dbplugin/v5/grpc_server.go [hunk 8]
```diff
@@ -16,19 +18,91 @@ var _ proto.DatabaseServer = gRPCServer{}
...
+
+	if db, ok := g.instances[id]; ok {
+		return db, nil
+	}
+
+	db, err := g.factoryFunc()
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (calvn) [inline sdk/database/dbplugin/v5/grpc_server.go:62]: If we changed factoryFunc to be `factoryFunc func() (interface{}, error)` (so that we authors can keep using New), this would be the probably be the place where you'd cast it back as a `Database`.

File: sdk/database/dbplugin/v5/grpc_server.go (hunk 5 above)
Comment (fairclothjm) [reply sdk/database/dbplugin/v5/grpc_server.go:69]: As it currently stands I don't think so since `getDatabase` is called by `Close()`. I think for Close we will want a Write lock so that connections to the database aren't closed during other operations (read, write, update).

Maybe it makes sense to have a separate getDatabase call that holds a write lock for Close.

File: sdk/database/dbplugin/v5/grpc_server.go (hunk 4 above)
Comment (fairclothjm) [reply sdk/database/dbplugin/v5/grpc_server.go]: addressed in: https://github.com/hashicorp/vault/pull/13734/commits/05b0db6d7b8cafca01493b19b143d4c4a8653f1a

File: sdk/database/dbplugin/v5/grpc_server.go (hunk 5 above)
Comment (fairclothjm) [reply sdk/database/dbplugin/v5/grpc_server.go]: addressed in https://github.com/hashicorp/vault/pull/13734/commits/05b0db6d7b8cafca01493b19b143d4c4a8653f1a

File: vault/plugin_catalog.go (hunk 6 above)
Comment (fairclothjm) [reply vault/plugin_catalog.go]: addressed in https://github.com/hashicorp/vault/pull/13734/commits/b9d185e11083094629bd6a60bdc37d7ed573edc2

File: plugins/database/postgresql/postgresql.go (hunk 7 above)
Comment (fairclothjm) [reply plugins/database/postgresql/postgresql.go]: addressed in https://github.com/hashicorp/vault/pull/13734/commits/e093c883f89a49d98395c796eeb9416961c33306

File: sdk/database/dbplugin/v5/grpc_server.go (hunk 8 above)
Comment (fairclothjm) [reply sdk/database/dbplugin/v5/grpc_server.go:62]: addressed in https://github.com/hashicorp/vault/pull/13734/commits/e093c883f89a49d98395c796eeb9416961c33306

File: sdk/database/dbplugin/v5/grpc_database_plugin.go [hunk 9]
```diff
@@ -6,20 +6,27 @@ import (
...
+const multiplexingCtxKey string = "multiplex_id"
+
+// Factory is the factory function to create a dbplugin Database.
+type Factory func() (interface{}, error)
+
 type GRPCDatabasePlugin struct {
-	Impl Database
+	FactoryFunc         Factory
+	Impl                Database
+	multiplexingSupport bool
```
Comment (fairclothjm) [inline sdk/database/dbplugin/v5/grpc_database_plugin.go:29]: This is not being used for anything at the moment. Maybe we don't need this?

File: sdk/database/dbplugin/v5/plugin_client.go [hunk 10]
```diff
@@ -3,59 +3,46 @@ package dbplugin
...
-
-	client, err := pluginRunner.RunConfig(ctx,
-		pluginutil.Runner(sys),
-		pluginutil.PluginSets(pluginSets),
-		pluginutil.HandshakeConfig(handshakeConfig),
-		pluginutil.Logger(logger),
-		pluginutil.MetadataMode(isMetadataMode),
-		pluginutil.AutoMTLS(true),
-	)
+	pluginClient, err := sys.NewPluginClient(ctx, pluginRunner, logger, isMetadataMode)
```
Comment (calvn) [inline sdk/database/dbplugin/v5/plugin_client.go:40]: How does `NewPluginClient` handle non-multiplexed client? `sys.NewPluginClient` seems to only return clients that are multiplexed.

File: vault/testing.go [hunk 11]
```diff
@@ -2148,7 +2148,7 @@ func (m *mockBuiltinRegistry) Get(name string, pluginType consts.PluginType) (fu
 		return nil, false
 	}
 	if name == "postgresql-database-plugin" {
-		return dbPostgres.New, true
+		return toFunc(dbPostgres.New), true
```
Comment (calvn) [inline vault/testing.go:2151]: I don't think this wrapping is necessary, you could `return dbPostgres.New, true`

**Review by calvn — COMMENTED** [review]
It's shaping up pretty well! Left a round of comments, but feel free to re-tag me for another look.

File: vault/testing.go (hunk 11 above)
Comment (fairclothjm) [reply vault/testing.go]: resolved in: https://github.com/hashicorp/vault/pull/13734/commits/10b2da3c1691b8c36418824b9cb75d9c9de82dc5

File: builtin/logical/database/path_config_connection.go [hunk 12]
```diff
@@ -365,6 +366,9 @@ func (b *databaseBackend) connectionWriteHandler() framework.OperationFunc {
 				"Vault (or the sdk if using a custom plugin) to gain password policy support", config.PluginName))
 		}
 
+		if resp.Warnings == nil {
```
[the remark this answers is not in the document]
Comment (fairclothjm) [reply builtin/logical/database/path_config_connection.go]: resolved in: https://github.com/hashicorp/vault/pull/13734/commits/10b2da3c1691b8c36418824b9cb75d9c9de82dc5

File: sdk/database/dbplugin/v5/grpc_server.go [hunk 13]
```diff
@@ -173,7 +271,15 @@ func (g gRPCServer) Type(ctx context.Context, _ *proto.Empty) (*proto.TypeRespon
...
+	impl, err := g.getDatabaseInternal(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	err = impl.Close()
 	if err != nil {
 		return &proto.Empty{}, status.Errorf(codes.Internal, "unable to close database plugin: %s", err)
 	}
```
Comment (calvn) [inline sdk/database/dbplugin/v5/grpc_server.go:285]: After closing, we should also deleting the instance from the `instances` map.

File: vault/testing.go (hunk 11 above)
Comment (fairclothjm) [reply vault/testing.go]: addressed in: https://github.com/hashicorp/vault/pull/13734/commits/10b2da3c1691b8c36418824b9cb75d9c9de82dc5

File: sdk/database/dbplugin/v5/grpc_server.go (hunk 13 above)
Comment (fairclothjm) [reply sdk/database/dbplugin/v5/grpc_server.go:285]: addressed in: https://github.com/hashicorp/vault/pull/13734/commits/e551043fb368569e1a7c643270854cacb3f41d48

File: sdk/helper/pluginutil/runner.go [hunk 14]
```diff
@@ -31,6 +33,13 @@ type LookRunnerUtil interface {
 	RunnerUtil
 }
 
+type Multiplexer interface {
+	ID() string
+	Conn() *grpc.ClientConn
+	Protocol() plugin.ClientProtocol
```
[the remark this answers is not in the document]
Comment (fairclothjm) [reply sdk/helper/pluginutil/runner.go]: addressed in: https://github.com/hashicorp/vault/pull/13734/commits/5f895a03477d29b1afb22969bd823847faf165af

File: vault/plugin_catalog.go (hunk 3 above)
Comment (fairclothjm) [reply vault/plugin_catalog.go]: addressed in: https://github.com/hashicorp/vault/pull/13734/commits/bee22a9ca5649da2dc01142094af018df8b430a3
